### ISSUE TITLE
Lazy enumerable select eagerly evaluates next item

### ISSUE THREAD

**Issue by headius** [jruby/jruby#4212]
Discovered by @twalpole in SeleniumHQ/selenium#2880. It appears our lazy enumerable logic is trying to evaluate the next item eagerly, or something...

``` ruby
class A
  include Enumerable

  def initialize()
    @result_enum = ["Text", "Text", "Text"].lazy.select do |t|
      puts "evaluating lazy select"
      true
    end
  end

  def each(&block)
    loop do
      block.call(@result_enum.next)
    end
  end
end

A.new().first
```

Running this on MRI prints "evaluating lazy select" once, while JRuby prints it twice.

**Comment by headius** [comment]
Printing out `caller(0)` shows that this logic is called twice with pretty much the same stack.

```
blah.rb:6:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:81:in `block in select'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:24:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:23:in `each'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:23:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:22:in `catch'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:22:in `block in initialize'
blah.rb:6:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:81:in `block in select'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:24:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:23:in `each'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:23:in `block in initialize'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:22:in `catch'
/Users/headius/projects/jruby/core/src/main/ruby/jruby/kernel/enumerator.rb:22:in `block in initialize'
```

**Comment by headius** [comment]
Reduced case; it's back to being a problem with `Enumerator#next` eagerly evaluating the next element:

```
enum = ["Text1", "Text2", "Text3"].lazy.select do |t|
  puts caller(0)
  true
end

enum.next
sleep
```

**Comment by headius** [comment]
This may be a good time to get rid of the native and separate fiber-like logic from Enumerator and just use Fiber.

**Comment by headius** [comment]
Ok, so here's the flaw in the design of our threaded `Enumerator#next`. In order to support `Enumerator#peek`, we eagerly try to get the next element. This causes the nexting thread to continue running after yielding a value, so it can acquire the next value, and as a result you have two threads operating against the same state at the same time.

Looking over MRI's implementation of `peek`, I can see they do this *lazily*. If you simply `next` your way through a collection, each yield stops the nexting thread (or Fiber in MRI's case). Only if you call `Enumerator#peek` does MRI attempt to grab the next value, which it then saves for any subsequent call to `Enumerator#next`.

This shouldn't be a difficult change to make. It might be less risk in a minor release than switching wholesale over to a Fiber-based `next`.

**Comment by headius** [comment]
Here's an implementation of Enumerator's next, peek, and rewind using Fibers and lazily acquiring the peek value. It works as expected for all above examples, but doesn't quite pass specs yet. It's probably close.

```diff
diff --git a/core/src/main/ruby/jruby/kernel/enumerator.rb b/core/src/main/ruby/jruby/kernel/enumerator.rb
index 4036949..d5cf978 100644
--- a/core/src/main/ruby/jruby/kernel/enumerator.rb
+++ b/core/src/main/ruby/jruby/kernel/enumerator.rb
@@ -1,4 +1,33 @@
 class Enumerator
+  UNDEF = Object.new
+
+  def next
+    if defined?(@__peek_value) && @__peek_value != UNDEF
+      ret, @__peek_value = @__peek_value, UNDEF
+      return ret
+    end
+
+    fiber = @fiber ||= Fiber.new do
+      each {|x| Fiber.yield x}
+    end
+
+    fiber.resume
+  end
+
+  def peek
+    if defined?(@__peek_value) && @__peek_value != UNDEF
+      return @__peek_value
+    end
+
+    @__peek_value = self.next
+  end
+
+  def rewind
+    @__peek_value = UNDEF
+    @fiber = nil
+    self
+  end
+
   class Yielder
     # Current API for Lazy Enumerator does not provide an easy way
     # to handle internal state. We "cheat" and use yielder to hold it for us.
```

Known issues:

* I believe this version ends up rooting the Enumerator and Fiber so they won't GC properly if abandoned. We need more isolation between the Fiber and the Enumerator.
* next_values, peek_values are not implemented, and various flow-control aspects of next and peek are missing (e.g. throwing StopIteration).
* Our shortcut logic to avoid threaded nexting on known collection types (like Array and Hash) is not wired in here. That would need to be reimplemented in Ruby alongside this logic.

Fiber has become reasonably stable in JRuby at this point, so having two separate Fiber

**Comment by kares** [comment]
najs, but we'll some how loose `Enumerator`'s recent `java.util.Iterator` behavior :(

**Comment by enebo** [comment]
@kares would it be possible to include java.util.Iterator and provide methods or will that no longer work?

**Comment by kares** [comment]
@enebo hopefully that would work but some state needs to be managed for `Iterator` (`hasNext` still needs to act as look-ahead) ... also thought there might be a collision with the `next` method (which I was wrong since I assumed it might receive *args)

**Comment by enebo** [comment]
@kares peek also does look-ahead but I guess the problem is we cannot tell if peek is returning nil from next or it really has no value?

@headius what happens if lazy is stored and called by two separate threads?  Isn't there a race for the fiber being created?

**Comment by headius** [comment]
@kares if RubyEnumerator still implements Iterator, then it should be fine. We'd only be moving a few methods from Java to Ruby and ditching the custom fiber logic in RubyEnumerator/Nexter.

Peek and hasNext would both trigger the next item to load. That's just how it has to work to satisfy both specs.

@enebo We spoke over IM but to close the loop: yes, this is a naive proof-of-concept and I have not filled out all features nor made any attempt to keep it thread-safe. I did this to show that a correct lazy impl does indeed fix the reported issue, and because reworking the Java-based `Enumerator#next` would take more work to be made lazy (but probably will be less risk).

**Comment by headius** [comment]
I looked over the Truffle impl of this (based on the Rubinius 2.x impl) and it also appears to have numerous race conditions.

**Comment by kares** [comment]
did not realize some of the `RubyEnumerator` parts are to be kept. sorry for the confusion on my part than.

**Comment by headius** [comment]
See also #4903.

**Comment by headius** [comment]
Marked #5007 as dupe.

**Comment by twalpole** [comment]
Any update on if or when this may get some attention? Would be nice to be able to remove JRuby specific stuff from the Capybara code base.

**Comment by headius** [comment]
@twalpole Could you test out #5672 and see if it works better? I still have some work to do there, but that's the direction. Once we can be assured that it doesn't leak fibers, it can get merged.

**Comment by twalpole** [comment]
@headius Sure -- Do you know if there an easy way to get rbenv to install that branch??  otherwise it's going to me a while to find the time to get the build stuff all set up and working.

**Comment by headius** [comment]
@twalpole Actually I took your comment as a call to action to finish the PR. I should have something landed on master shortly.

**Comment by headius** [comment]
FWIW the new logic works properly.

**Comment by twalpole** [comment]
@headius Ok - I built the distribution files from your branch and then installed it into my rbenv - and now I get a bunch of errors like - 

    An error occurred while loading ./spec/session_spec.rb.
Failure/Error: # frozen_string_literal: true

```
LoadError:
  load error: childprocess/jruby -- java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.io.FileDescriptor sun.nio.ch.FileChannelImpl.fd accessible: module java.base does not "opens sun.nio.ch" to module org.jruby.dist
# ./spec/spec_helper.rb:1:in `block in (root)'
# ./spec/spec_helper.rb:6:in `<main>'
# ./spec/session_spec.rb:1:in `(root)'
# ./spec/session_spec.rb:3:in `<main>
```

When attempting to run the Capybara specs -- so not sure if it's my building/installing (probably) or your branch  that has issues - I'll wait for it to go into master and then try again

**Comment by headius** [comment]
That looks like a Java 9 issue...and also an issue that was fixed in the release but probably not on this branch. Hold off and I'll merge to master shortly.

**Comment by headius** [comment]
I've merged master back into #5672 and just waiting to make sure CI looks good.

**Comment by headius** [comment]
Fixed by #5672.

@twalpole Could you come up with some simple specs that test this behavior? You can submit as a PR to JRuby or https://github.com/ruby/spec under spec/ruby/core/enumerator.

**Comment by twalpole** [comment]
Ok - so using the master branch with Java 8 (can't use 9+ because the current release of cucumber doesn't support child_process 1.0.x which is required for Java 9+...) this problem is fixed - but a bunch of new stuff in the Capybara test suite breaks.  I'm going to have to dig deeper into it tomorrow, and will try and get some simple tests for the lazy enumerator behavior submitted tomorrow too.
